docs/proposals/improve-status-in-CAPI-resources.md

```diff
@@ -0,0 +1,921 @@
+To improve the benefit of this change:
+
+- This proposal is ensuring that whenever Machine ready is used, it always means the same thing (e.g. replica counters)
+- This proposal is also changing contract fields where ready was used improperly to represent 
+  initial provisioning (k8s API conventions suggest to use ready only for long-running process).
+
+All in all, Machine's Ready should be much more clear, consistent, intuitive after proposed changes.
+But there is more.
+
+This proposal is also dropping Ready condition from higher level abstractions in Cluster API.
+
+Instead, where not already present, this proposal is introducing a new Available condition that better represents
+the fact that those objects are operational even if there is a certain degree of not readiness in the system
+or if lifecycle operations are happening (prior art Available condition in K8s Deployments).
+
+Last but not least:
+
+- With the changes to the semantic of Ready and Available conditions, it is now possible to add conditions about
+  surfacing that lifecycle operations are happening, e.g. scaling up.
+- As suggested by K8s API conventions, this proposal is also making sure all conditions are consistent and have
+  uniform meaning across all resource types
+- Additionally, we are enforcing the same consistency for replica counters and other status fields.
+
+### Transition to K8s API conventions aligned conditions
+
+K8s is undergoing an effort of standardizing usage of conditions across all resource types, and the transition to
+the v1beta2 API version is a great opportunity for Cluster API to align to this effort.
+
+The value of this transition is substantial, because the differences that exists today's are really confusing for users;
+those differences are also making it harder for ecosystem tools to build on top of Cluster API, and in some cases
+even confusing new (and old) contributors.
+
+With this proposal Cluster API will close the gap with K8s API conventions in regard to:
+- Polarity: Condition type names should make sense for humans; neither positive nor negative polarity can be recommended
+  as a general rule (already implemented by [#10550](https://github.com/kubernetes-sigs/cluster-api/pull/10550))
+- Use of the Reason field is required (currently in Cluster API reasons is added only when condition are false)
+- Controllers should apply their conditions to a resource the first time they visit the resource, even if the status is Unknown.
+  (currently Cluster API controllers add conditions at different stages of the reconcile loops)
+- Cluster API is also dropping its own Condition type and start using metav1.Conditions from the Kubernetes API.
+
+The last point have also another implication, which is the removal of the Severity field which is currently used
+to determine priority when merging conditions.
+
+TODO Document how we are going to replace severity (currently prototyping)
```

Comment:
One example I've had in mind was something like: 
- Say I'm creating ec2 instances
- Try to create instance, but waiting for some other component (like IPAM) to allocate address
- A `InstanceCreated` condition could be false, in warning state, and explain why.

One could argue we don't really need to distinguish between "giving a warning" vs "error/failure" and that's enough for that condition to just report false; which might be ok after all?

Reply:
I finally had time to write this up, PTAL
TL;DR; I don't think that we will need this because with the revised concept or readyness it is probably better to surface all the messages vs arbitrarily hiding some (and merge priority to pick the single message to surface when computing ready).

But I think that our conditions utils should make it possible for developers to plug in custom func to determine merge priority for conditions, if the need this for some use case (conditions util were used broadly in providers, and now that we are aligning them to K8s, they can potentially be used in more use cases).

Reply:
I like the idea of having an ability to customise the priority. If X condition and Y condition are in their non-normative state, leaving it to the provider to suggest which one is a priority is very flexible and less likely to cause issues if we make blanket assumptions, SGTM

Reply:
Agree, all good as long as it is a "local" decision by the provider (or in general the controller owning the conditions / resource). I think we should just not try to "hand over" that priority information to the core CAPI controller

Reply:
It seems there is consensus, resolving